gpkit/constraints/sgp.py

```diff
@@ -184,44 +182,38 @@ def firstgp(self, x0, substitutions):
                 else:
                     self.is_sgp = True
                     return
-        self._gppos = 1 + len(gpposys)
-        return [gpconstrs,
-                [p/m <= 1 for p, m in zip(self._approx_lt, approx_gt)]]
-
-    def gp(self, x0=None, verbosity=1, modifylastgp=False):
+        spapproxs = [p/m <= 1 for p, m in zip(self._approx_lt, approx_gt)]
+        gp = GeometricProgram(self.cost, [gpconstrs, spapproxs],
+                              substitutions, verbosity=verbosity)
+        gp.x0 = x0
+        self._gppos = len(gp.hmaps) - len(spapproxs)
+        return gp
+
+    def gp(self, x0=None, verbosity=1, mutategp=False):
         "The GP approximation of this SP at x0."
-        x0 = self._fill_x0(x0)
-        if not hasattr(self, "externalfn_vars"):
-            self.__parse_externalfnvars()
-        if modifylastgp and self.lastgp:
-            spmonos = []
-            for spc in self._spconstrs:
-                spmonos.extend(spc.as_approxsgt(x0))
-            for i, spmono in enumerate(spmonos):
-                firstposy = self._approx_lt[i]
-                unsubbed = firstposy/spmono
-                self.lastgp[0][1][i].unsubbed = [unsubbed]
-                localposylt1 = unsubbed.sub(self.substitutions)
-                self.lastgp.posynomials[self._gppos+i] = localposylt1
-            self.lastgp.gen()
-            return self.lastgp
+        if mutategp and not self.is_sgp:
+            if self.gps:  # update self._gp with new x0
+                self._gp.x0.update(x0)
+                spmonos = []
+                for spc in self._spconstrs:
+                    spmonos.extend(spc.as_approxsgt(self._gp.x0))
```

Comment:
It's not quite that for SPEquality constraints, but I like that name a lot better anyway, done.